src/measure_theory/group/geometry_of_numbers.lean

```diff
@@ -0,0 +1,285 @@
+open_locale pointwise
+
+namespace measure
+variables {E : Type*} [normed_add_comm_group E] [measurable_space E] [normed_space ℝ E]
+  [finite_dimensional ℝ E] [borel_space E] (μ : measure_theory.measure E) [is_add_haar_measure μ]
+
+lemma add_haar_smul_of_nonneg {r : ℝ} (hr : 0 ≤ r) (s : set E) :
+  μ (r • s) = ennreal.of_real (r ^ finrank ℝ E) * μ s :=
+by rw [add_haar_smul, abs_pow, abs_of_nonneg hr]
+
+end measure
+
+section
+variables {𝕜 G H : Type*} [nontrivially_normed_field 𝕜] [complete_space 𝕜] [measurable_space G]
+  [topological_space G] [add_comm_group G] [module 𝕜 G] [finite_dimensional 𝕜 G]
+  [has_continuous_smul 𝕜 G] (μ : measure G) [is_add_haar_measure μ] [borel_space G] [t2_space G]
+  [topological_add_group G] [topological_space H] [add_comm_group H] [module 𝕜 H]
+  [finite_dimensional 𝕜 H] [has_continuous_smul 𝕜 H] [measurable_space H] [borel_space H]
+  [t2_space H] [topological_add_group H]
+
+instance (e : G ≃ₗ[𝕜] H) : is_add_haar_measure (μ.map e) :=
+e.to_add_equiv.is_add_haar_measure_map _ (e : G →ₗ[𝕜] H).continuous_of_finite_dimensional
+  (e.symm : H →ₗ[𝕜] G).continuous_of_finite_dimensional
+
+end
+
+section
+variables {α G V : Type*} [measurable_space G] [measurable_space α] {μ : measure G}
+
+@[to_additive]
+instance is_mul_left_invariant.to_smul_invariant_measure [has_mul G] [has_measurable_mul G]
+  [μ.is_mul_left_invariant] : smul_invariant_measure G G μ :=
+⟨λ g s hs, by simp_rw [smul_eq_mul, ←map_apply (measurable_const_mul g) hs, map_mul_left_eq_self]⟩
+open smul_invariant_measure
+
+variables [group G]
+
+@[to_additive] instance subgroup.smul_invariant_measure [has_measurable_mul G]
+  [μ.is_mul_left_invariant] {Γ : subgroup G} : smul_invariant_measure Γ G μ :=
+⟨λ c s hs, measure_preimage_mul μ _ s⟩
+
+@[to_additive] instance subgroup.smul_invariant_measure_op [has_measurable_mul G]
+  [μ.is_mul_right_invariant] {Γ : subgroup G} : smul_invariant_measure Γ.opposite G μ :=
+⟨λ c s hs, measure_preimage_mul_right μ _ s⟩
+
+end
+
+lemma rescale (ι : Type*) [fintype ι] {r : ℝ} (hr : 0 < r) :
+  comap ((•) r) (volume : measure (ι → ℝ)) = ennreal.of_real r ^ card ι • volume :=
+begin
+  suffices : (ennreal.of_real r ^ card ι)⁻¹ • comap ((•) r) (volume : measure (ι → ℝ)) = volume,
+  { conv_rhs { rw ←this },
+    rw [ennreal.inv_pow, smul_smul, ←mul_pow, ennreal.mul_inv_cancel (ennreal.of_real_pos.2 hr).ne'
+      ennreal.of_real_ne_top, one_pow, one_smul] },
+  refine (pi_eq $ λ s hS, _).symm,
+  simp only [smul_eq_mul, measure.coe_smul, pi.smul_apply],
+  rw [comap_apply _ (smul_right_injective (ι → ℝ) hr.ne') (λ S hS, hS.const_smul₀ r) _
+    (measurable_set.univ_pi hS), image_smul, smul_univ_pi, volume_pi_pi],
+  simp only [add_haar_smul, finite_dimensional.finrank_self, pow_one, abs_of_pos hr, pi.smul_apply,
+    finset.prod_mul_distrib, finset.card_univ, ←mul_assoc, finset.prod_const],
+  rw [ennreal.inv_mul_cancel _ (ennreal.pow_ne_top ennreal.of_real_ne_top), one_mul],
+  positivity,
+end
```

Comment:
@sgouezel, do you have a quicker route to this fact?

Reply:
I'm surprised it takes this much work to massage this into the statement of `add_haar_smul`. Are there lemmas you can refactor out? Maybe `comap_smul_eq`? Or `map_smul_add_haar`?

Reply:
I did not manage. If you have more precise suggestions, I can try again